bee-protocol/src/workers/peer/packet_handler.rs

```diff
@@ -0,0 +1,333 @@
+pub(super) struct PacketHandler {
+    events: EventHandler,
+    // FIXME: see if we can implement `Stream` for the `PacketHandler` and use the
+    // `ShutdownStream` type instead.
+    shutdown: ShutdownRecv,
+    state: ReadState,
+    /// The address of the peer. This field is only here for logging purposes.
+    address: Multiaddr,
+}
+
+impl PacketHandler {
+    /// Create a new packet handler from an event receiver, a shutdown receiver and the peer's
+    /// address.
+    pub(super) fn new(receiver: EventRecv, shutdown: ShutdownRecv, address: Multiaddr) -> Self {
+        Self {
+            events: EventHandler::new(receiver),
+            shutdown,
+            // The handler should read a header first.
+            state: ReadState::Header,
+            address,
+        }
+    }
+    /// Fetch the header and payload of a packet.
+    ///
+    /// This method only returns `None` if a shutdown signal is received.
+    pub(super) async fn fetch_packet(&mut self) -> Option<(HeaderPacket, &[u8])> {
+        // loop until we can return the header and payload
+        loop {
+            match &self.state {
+                // Read a header.
+                ReadState::Header => {
+                    // We need `HEADER_SIZE` bytes to read a header.
+                    let bytes = self
+                        .events
+                        .fetch_bytes_or_shutdown(&mut self.shutdown, HEADER_SIZE)
+                        .await?;
+                    trace!("[{}] Reading Header...", self.address);
+                    let header = HeaderPacket::from_bytes(bytes);
+                    // Now we are ready to read a payload.
+                    self.state = ReadState::Payload(header);
+                }
+                // Read a payload.
+                ReadState::Payload(header) => {
+                    // We read the quantity of bytes stated by the header.
+                    let bytes = self
+                        .events
+                        .fetch_bytes_or_shutdown(&mut self.shutdown, header.packet_length.into())
+                        .await?;
+                    // FIXME: Avoid this clone
+                    let header = header.clone();
+                    // Now we are ready to read the next packet's header.
+                    self.state = ReadState::Header;
+                    // We return the current packet's header and payload.
+                    return Some((header, bytes));
+                }
+            }
+        }
+    }
+}
+
+// An event handler.
+//
+// This type takes care of actually receiving the events and appending them to an inner buffer so
+// they can be used seamlessly by the `PacketHandler`.
+struct EventHandler {
+    receiver: EventRecv,
+    buffer: Vec<u8>,
+    offset: usize,
+}
+
+impl EventHandler {
+    /// Create a new event handler from an event receiver.
+    fn new(receiver: EventRecv) -> Self {
```

Comment:
`#[derive(Default)]`?

Reply:
What do you mean ? Can't derive it because of EventRecv ?

Reply:
I'm an idiot